3단계 검증요청

kht2199 opened the pull request:
안녕하세요. 조금 늦었네요..

수정사항을 아래에 정리했습니다.

1. 도메인 영역에서 요구사항에 해당하는 테스트 케이스 작성.
2. 기존 테스트 검증 통과. 하지만 delete시 answer를 따로 조회하지 않고, 삭제를 하도록 구현되어 기존 테스트의 answer를 조회하는 코드를 수정/삭제.
3. AnswerList로 일급컬렉션을 적용.
4. CreationTimestamp, UpdateTimestamp 적용

Commit:
test: Test 항목에 대한 DisplayName 적용

Commit:
fix: User의 userId를 unique로 변경

Commit:
fix: Entity 간의 연관맵핑 추가

Commit:
feat: Foreign keys to Relational mappings

Commit:
feat: Question 에서 Answer 연관관계 설정 추가

Commit:
feat: QnaService#deleteQuestion 도메인 용어로 구현

Commit:
test: 질문 데이터를 완전히 삭제하는 것이 아니라 데이터의 상태를 삭제 상태(deleted - boolean type)로 변경한다.

Commit:
test: 로그인 사용자와 질문한 사람이 같은 경우 삭제할 수 있다.

Commit:
doc: 질문 삭제가능한 경우 케이스 정리.

Commit:
feat: CreationTimestamp, UpdateTimestamp 적용

Commit:
feat: 질문 전체 삭제 기능 추가.

Commit:
test: 질문자와 답변 글의 모든 답변자 같은 경우 삭제가 가능하다

Commit:
doc: 테스트 완료된 케이스 상태 업데이트

Commit:
chore: 임시 저장

Commit:
test: answer repository를 사용하지 않으므로, 조회 로직을 삭제.

Commit:
test: 질문과 답변 삭제 이력에 대한 정보를 DeleteHistory를 활용해 남긴다. (Service test)

Commit:
Merge branch 'kht2199' of github.com:next-step/jwp-qna into kht2199

 Conflicts:
	src/main/java/qna/domain/DeleteHistory.java
	src/main/java/qna/domain/Question.java
	src/main/java/qna/domain/question/Answer.java
	src/main/java/qna/service/QnaService.java
	src/test/java/qna/service/QnaServiceTest.java

Commit:
fix: Audit 변경 - hibernate의 Timestamp를 spring Date로 변경

parkeeseul commented on src/main/java/qna/service/QnaService.java:
```diff
@@ -36,24 +39,19 @@ public Question findQuestionById(Long id) {
-            deleteHistories.add(new DeleteHistory(ContentType.ANSWER, answer.getId(), answer.writer()));
-        }
-        deleteHistoryService.saveAll(deleteHistories);
+        questionRepository.delete(question);
```
해당 미션에서는 디비에 값을 삭제하는 hard delete 가 아닌 soft delete 를 사용하고 있습니다! 😄

parkeeseul commented on src/main/java/qna/service/QnaService.java:
```diff
@@ -36,24 +39,19 @@ public Question findQuestionById(Long id) {
         Question question = findQuestionById(questionId);
-        if (!question.isOwner(loginUser)) {
-            throw new CannotDeleteException("질문을 삭제할 권한이 없습니다.");
+        AnswerList deletedAnswers;
+        try {
+            deletedAnswers = question.deleteBy(loginUser);
+        } catch (QuestionOwnerNotMatchedException e) {
+            throw new CannotDeleteException("질문을 삭제할 권한이 없습니다.", e);
+        } catch (AnswerOwnerNotMatchedException e) {
+            throw new CannotDeleteException("다른 사람이 쓴 답변이 있어 삭제할 수 없습니다.", e);
         }
```
Exception 를 해당 부분에서 핸들링 하는 이유가 있으실까요?
QuestionOwnerNotMatchedException 를 throw 하는 곳에서 CannotDeleteException 을 throw 해주면 어떨까요? 😄

kht2199 replied:
Domain과 Service 발생하는 Exception을 분리해서 적용한 이유입니다.
1. Domain에서 발생할 수 있는 다양한 Exception을 Service에서 다르게 처리할 수 있도록 함
2. Domain에서 발생할 수 있는 예외사항을 Exception 이름을 통해서 알수 있도록 하기 위함
3. 실제 업무에서 도메인 영역을 Service와 의존성이 분리해서 개발중입니다. 그래서, Service에서 정의한 Exception이나 다른 클래스를 사용하지 못하도록 하고있습니다. 그래서 이번 예제에서도, 따로 정의해서 사용하도록 했습니다.

parkeeseul commented on src/main/java/qna/service/QnaService.java:
```diff
@@ -36,24 +39,19 @@ public Question findQuestionById(Long id) {
-            throw new CannotDeleteException("질문을 삭제할 권한이 없습니다.");
+        AnswerList deletedAnswers;
+        try {
+            deletedAnswers = question.deleteBy(loginUser);
```
Question 을 삭제할때 question 과 answer 에 대한 deleteHistories 반환해보면 어떨까요? 😄

kht2199 replied:
다음 PR에 수정적용했습니다 ^^

parkeeseul commented on src/main/java/qna/domain/history/DeleteHistoryList.java:
```diff
@@ -0,0 +1,32 @@
+ *
+ * @author heetaek.kim
+ */
+public class DeleteHistoryList {
```
List 인 histories 의 타입이 변경된다면 해당 클래스의 네이밍 또한 변경되어야 한다는 이슈가 있을 것 같아요!
DeleteHistories 라는 네이밍은 어떨까요? 😄

kht2199 replied:
네이밍 변경했습니다. 로또  때에는 모두 List로 명명했었는데, 생각해보니 복수형이 더 보기가 좋기도 하고 변경도 용이하겠네요!

parkeeseul commented on src/main/java/qna/domain/question/Answer.java:
```diff
@@ -58,18 +62,25 @@ public boolean isDeleted() {
+            '}';
+    }
+
+    public void deleteBy(User loginUser) throws AnswerOwnerNotMatchedException {
+        if (!isOwner(loginUser)) {
+            throw new AnswerOwnerNotMatchedException();
+        }
+        delete();
+    }
+
+    private void delete() {
+        this.deleted = true;
     }
```
클래스의 구현 순서는 다음과 같이 중요한 것 부터 덜 중요한 순으로 위치시키는 것이 좋은데요
한번 확인 해보시면 좋을 것 같아요! 😄 
``` java
class A {
    상수(static final) 또는 클래스 변수
    인스턴스 변수
    생성자
    팩토리 메소드
    메소드
    기본 메소드 (equals, hashCode, toString)
}
```

kht2199 replied:
다음 PR에서 순서 변경해서 적용했습니다

parkeeseul commented on src/main/java/qna/domain/question/AnswerList.java:
```diff
@@ -0,0 +1,43 @@
+ *
+ * @author heetaek.kim
+ */
+public final class AnswerList {
```
해당 클래스의 네이밍도 한번 확인 부탁드릴게요! 😄

kht2199 replied:
어떤 네이밍이 좋을까요? 일급 컬렉션을 적용할때, 리스트는 List를 붙이고 있습니다.

parkeeseul commented on src/main/java/qna/domain/question/AnswerList.java:
```diff
@@ -0,0 +1,43 @@
+			answer.deleteBy(deleteBy);
+			deleted.add(answer);
+		}
+		answers.removeAll(deleted);
```
removeAll을 하는 이유가 있으실까요? 🤔

kht2199 replied:
AnswerList의 list는 Answer의 list를 참조하고 있습니다,  그래서 orphan removal 을 true로 적용하고, 리스트에서 제거될 경우 삭제될 수 있도록 하려고 했습니다

kht2199 replied:
제가 착각했네요. ㅜㅜ orphanRemoval을 적용해서 삭제하도록 하려고 했었는데, 논리삭제라 실제 삭제가 되면 안되네요.

removeAll, removal을 모두 제거했습니다.

parkeeseul commented on src/main/java/qna/domain/question/AnswerList.java:
```diff
@@ -0,0 +1,43 @@
+	 * @return 삭제된 질문 목록.
+	 * @throws AnswerOwnerNotMatchedException 삭제하는 사용자가 권한이 없을 경우 발생.
+	 */
+	public AnswerList deleteAllBy(User deleteBy) throws AnswerOwnerNotMatchedException {
+		List<Answer> deleted = new ArrayList<>();
+		for (Answer answer : answers) {
+			answer.deleteBy(deleteBy);
+			deleted.add(answer);
+		}
+		answers.removeAll(deleted);
+		return new AnswerList(deleted);
+	}
```
해당 메서드에서 DeleteHIstory 를 반환하는 것은 어떻게 생각하시나요? 😄

kht2199 replied:
다음 PR에서 적용되었습니다.

parkeeseul commented on src/test/java/qna/domain/question/QuestionTest.java:
```diff
@@ -0,0 +1,90 @@
+        }
+        assertThatThrownBy(() -> question.deleteBy(loginUser))
+            .isInstanceOf(QuestionOwnerNotMatchedException.class);
+    }
```
질문을 삭제 할 수 있는 경우와 삭제 할 수 없는 경우에 대한 테스트가 혼재 하고 있네요!
테스트 코드 또한 가독성이 중요한데요 한 테스트에 두가지의 개념이 들어가게 되면 독자는 한번에 두가지의 개념을 이해해야 하기 때문에 가독성이 떨어 질 수 있을 것 같아요!
테스트 코드 하나당 한 개념만 테스트 하도록 수정해주면 어떨까요? 😄

parkeeseul reviewed (changes requested):
리팩토링 잘 진행해주셨네요! 👍 
몇 가지 코멘트 남겨드렸으니 확인 부탁드립니다! 😄 
미션 화이팅 입니다 ! 🔥

kht2199 committed:
fix: Question 삭제시, AnswerList가 아닌 void를 반환하도록 수정
1. 삭제된 Question을 반환하는게 맞지만, 굳이 반환할 필요가 없어서 반환하지 않음.
2. question history 등록시, AnswerList를 question에서 가져올 수 있도록 getter 제공.

kht2199 committed:
fix: Question 삭제시, DeleteHistoryList를 반환하도록 수정.

kht2199 committed:
fix: Answer toString method reorder

kht2199 committed:
fix: DeleteHistoryList에서 Question과 Answer를 분리

kht2199 committed:
refactor: List 명명을 복수형으로 변경, List를 사용하는 부분에서는 가능한 Collection으로 타입 변경.

kht2199 committed:
test: 질문 삭제 케이스 분리, 공통코드 Each로 이동.

parkeeseul commented on src/main/java/qna/domain/exception/question/QuestionNotDeletedException.java:
```diff
@@ -0,0 +1,13 @@
+import qna.domain.exception.DomainException;
+import qna.domain.question.Question;
+
+public class QuestionNotDeletedException extends DomainException {
+
+	private final Question question;
+
+	public QuestionNotDeletedException(Question question) {
+		this.question = question;
+	}
+}
```
question 을 상태값으로 갖는 이유가 있으신가요? 🤔

kht2199 replied:
아뇨 ㅎ 의미는 없어요. 나중에 question 정보 중에, 에외처리시 필요한게 있으면 사용할 수 있도록 레퍼런스를 넣어뒀는데, 굳이 필요는 없을것같아요.

kht2199 replied:
Message로 수정했습니다.

parkeeseul commented on src/main/java/qna/domain/question/Question.java:
```diff
@@ -0,0 +1,108 @@
+    @JoinColumn(foreignKey = @ForeignKey(name = "fk_question_writer"))
+    private User writer;
+
+    @OneToMany(mappedBy = "question", cascade = CascadeType.REMOVE)
+    private final List<Answer> answers = new ArrayList<>();
+
+    @Transient
+    private final Answers answersWrapped = new Answers(answers);
```
질문에 대한 답변을 추가하게 되면 answers 와 answersWrapped 를 모두 관리해줘야 하는 이슈가 있을 것 같아요
@Embedded 를 활용해보면 어떨까요? 😄 
```suggestion
    @Embedded
    private final Answers answers = new Answers();
```

kht2199 replied:
그렇군요~ 그런 방법을 잊고있었네요. 다음 PR때 적용하도록 할게요.

parkeeseul reviewed (changes requested):
피드백 잘 반영해주셨네요! 👍 
몇 가지 추가 코멘트 남겨드렸으니 확인 부탁드립니다! 😄 
주말은 왜이리 시간이 빠른지 모르겠어요! 남은 하루 잘 마무리 하시길 바라요!
미션 화이팅 입니다! 🔥

kht2199 committed:
refactor: Answers embeddable로 변경.

kht2199 committed:
fix: List, Collection type mismatch.

kht2199 committed:
fix: Exception에서 메시지 등록

parkeeseul commented on src/main/java/qna/domain/question/Question.java:
```diff
@@ -87,11 +79,8 @@ public Answer addAnswer(Answer answer) {
     public Answer addAnswer(User answerUser, String content) {
         Answer answer = new Answer(answerUser, this, content);
-        return addAnswer(answer);
-    }
```
해당 메서드는 테스트 코드에서만 사용하고 있네요! 🤔 
테스트 코드를 위한 비즈니스 로직은 지양해야한다고 생각하는데 제거해주는게 어떨까요? 😄

parkeeseul reviewed (approved):
피드백 잘 반영해주셨네요! 👍 
소소한 마지막 코멘트 남겨두었으니 한번 확인 부탁드리고
이만 머지 하도록 하겠습니다! 😄 
그동한 많은 피드백 반영해주시느라 고생 많으셨습니다!
다소 많은 피드백 드렸었는데 잘 반영해주셔서 감사해요! 🙇‍♀️ 
남은 미션들도 꼭 완주하셔서 많은 인사이트 얻어가시길 바라겠습니다!
마지막까지 화이팅 입니다! 🔥
